### DISCUSSION TITLE
Pseudo data -- fossil ages, etc.

### DISCUSSION THREAD

**Discussion by bredelings** [revbayes/revbayes#622] [General]
I've also been thinking about pseudodata, which came up again in putting ages on FBD taxa like
```
    F[i] ~ dnUniform(t[i] - b[i], t[i] - a[i])
    F[i].clamp( 0 )
```
In theory we could do something like:
```
   require( b[i] < t[i] && t[i] < a[i] )
```
But that is not general -- it wouldn't handle log-normals or intervals with soft bounds.

So I'm thinking something like:
```
   F[i] ~ dnPseudoData(t[i])
   F[i].clamp( pdInterval(a[i], b[i]) )
```

And then if you want soft bounds on the interval with exponential decay rate lambda we could have:

```
   F[i] ~ dnPseudoData(t[i])
   F[i].clamp( pdSoftBounds(a[i], b[i], lambda) )
```

Note that technically `pdInterval(a[i],b[i])` is different that `dnUniform(a[i],b[i])` because:
- `dnUniform(a[i],b[i])` has density `1/(b[i] - a[i])` inside the interval and `0` outside
- `pdInterval(a[i],b[i])` has likelihood `1` inside the interval, and `0` outside.

The idea is that `pdSoftBounds(a,b,lambda)` would have:
- density `1` inside the interval.
- density `exp(-lambda*(a-x))` below the interval.
- density `exp(-lambda*(b-x))` above the interval.

**Comment by bredelings** [comment]
I guess `dnPseudoData(x)` could have type `PseudoDataOnReal` if `x` is `Real`, `PseudoDataOnInteger` if `x` is `Integer`, etc.

**Comment by davidcerny** [comment]
> I guess `dnPseudoData(x)` could have type `PseudoDataOnReal` if `x` is `Real`, `PseudoDataOnInteger` if `x` is `Integer`, etc.

This is inconsistent with how we currently do things, but I love it. :) Some of our unresolved issues stem from the fact that instead of having `dnUniformOnReal`, `dnUniformOnProbability`, etc., we just try to guess what the return type of `dnUniform` should be based on its bounds. Not an unreasonable thing to do perhaps – if there was a way to override the guess!
